### PR TITLE
[WS-31] Mocking WalletService

### PR DESCRIPTION
**Pull request by mfaulk**
Soundtrack of this PR: [Max Cooper - Repetition](https://youtu.be/nO9aot9RgQc)

### Motivation

Mock objects make it easy to unit test individual components without laboriously constructing and initializing real dependencies. This is particularly handy for testing API-layer functionality.

### In this PR
[x] Extracts a Wallet trait,
[x] Uses mockall to generate MockWallet,
[x] Updates WalletState to hold an instance of Wallet,
[x] Updates wallet API unit tests to use mocked WalletState,

This PR is smaller than it looks because it contains some refactors:
- Splits error.rs into separate files so that the errors are defined closer to the code that generates them.
- Moves "API-related" things into a new api module (e.g. wallet.rs --> api/wallet.rs).
- Moves the large enums JsonCommandRequest and JsonCommandResponse into separate files (api/json_rpc_request.rs and api/json_rpc_response).
- Splits the JSON-catch-all file decorated_types.rs into separate files under the `api` module.

### Future Work
* Move/Add unit tests that test the wallet service independent of the API.
* Try to encapsulate JSON and String-y data into a JSON-RPC API Layer, so that the application layer (i.e. WalletService) works with normal Rust objects. That would make developing and testing the application layer easier, and would also make it easy to add other API implementations (gRPC, command line, GUI).

### CODE REVIEW

File: full-service/src/db/wallet_db_error.rs [hunk 1]
```diff
@@ -0,0 +1,112 @@
+use displaydoc::Display;
+
+#[derive(Display, Debug)]
+pub enum WalletDbError {
```
Comment (mfaulk) [inline full-service/src/db/wallet_db_error.rs:4]: I moved several error types out of the top-level error.rs and into new files that are closer to where the error is generated. I think this helps organize the code more explicitly around API / Application Logic / Database layers.

File: full-service/src/service/sync_error.rs [hunk 2]
```diff
@@ -0,0 +1,62 @@
+use displaydoc::Display;
+use crate::db::WalletDbError;
+
+#[derive(Display, Debug)]
+pub enum SyncError {
```
Comment (mfaulk) [inline full-service/src/service/sync_error.rs:5]: This is not new, it was just moved.

File: full-service/src/service/transaction_builder_error.rs [hunk 3]
```diff
@@ -0,0 +1,95 @@
+use displaydoc::Display;
+use crate::db::WalletDbError;
+
+#[derive(Display, Debug)]
+pub enum WalletTransactionBuilderError {
```
Comment (mfaulk) [inline full-service/src/service/transaction_builder_error.rs:5]: This is not new, it was just moved.

File: full-service/src/service/wallet.rs [hunk 4]
```diff
@@ -1,1067 +0,0 @@
-// Copyright (c) 2020-2021 MobileCoin Inc.
```
Comment (mfaulk) [inline full-service/src/service/wallet.rs:1]: This file became wallet_api.rs

File: full-service/src/service/wallet_api.rs [hunk 5]
```diff
@@ -0,0 +1,1119 @@
...
+        let result = dispatch(&client, body, &logger);
+        assert!(result.get("entropy").is_some());
+        assert_eq!(result.get("entropy").unwrap(), expected_response.entropy.as_str());
+
+        let account_obj = result.get("account").unwrap();
+        assert_eq!(account_obj.get("account_id").unwrap(), expected_response.account.account_id.as_str());
+    }
+
+    // #[test_with_logger]
+    // fn test_account_crud(logger: Logger) {
```
Comment (mfaulk) [inline full-service/src/service/wallet_api.rs:664]: These unit tests seem to focus on checking the implementation of the WalletService, and so the "meat" of these tests should probably get moved to wallet_service.rs

File: full-service/src/service/wallet_service.rs [hunk 6]
```diff
@@ -101,9 +103,68 @@ impl<
             logger,
         }
     }
+}
+
+impl<
+    T: BlockchainConnection + UserTxConnection + 'static,
+    FPR: FogPubkeyResolver + Send + Sync + 'static,
+> Wallet for WalletService<T, FPR> {
```
Comment (mfaulk) [inline full-service/src/service/wallet_service.rs:111]: I don't think any of this code is new. It just moved around because these functions are now declared in the Wallet trait.

File: full-service/src/service/wallet_service_error.rs [hunk 7]
```diff
@@ -0,0 +1,114 @@
+use displaydoc::Display;
+use crate::db::WalletDbError;
+use crate::service::transaction_builder_error::WalletTransactionBuilderError;
+
+#[derive(Display, Debug)]
+pub enum WalletServiceError {
```
Comment (mfaulk) [inline full-service/src/service/wallet_service_error.rs:6]: This isn't new, it just moved.

File: full-service/src/service/wallet_trait.rs [hunk 8]
```diff
@@ -0,0 +1,276 @@
...
+use crate::db::account::AccountID;
+use mc_mobilecoind_json::data_types::{JsonTxProposal, JsonTx, JsonTxOut};
+
+#[cfg(test)]
+use mockall::*;
+
+
+/// A MobileCoin wallet.
+#[cfg_attr(test, automock)]
+pub trait Wallet {
```
Comment (mfaulk) [inline full-service/src/service/wallet_trait.rs:12]: This is new.

I'm working on filling in the ??? placeholders in the doc comments, but suggested text is welcome!

File: full-service/src/lib.rs [hunk 9]
```diff
@@ -4,13 +4,14 @@
 // Required because hashbrown is at 0.9.1 and has build issues otherwise
 #![feature(ptr_offset_from)]
 
+mod api;
```
Comment (mfaulk) [inline full-service/src/lib.rs:7]: This module is to help organize the "API layer" of the application. I think this will be helpful if/when we want to support other API modes (e.g. gRPC), and should help encapsulate where JSON and String-y data are used. I also just like the familiar feel of web applications that are organized around API / Application Logic (Services) / Database (Persistence).

File: full-service/src/json_rpc/mod.rs [hunk 10]
```diff
@@ -0,0 +1,28 @@
+// Copyright (c) 2020-2021 MobileCoin Inc.
+
+//! JSON-RPC-2.0 API
```
Comment (mfaulk) [inline full-service/src/json_rpc/mod.rs:3]: Most of the code here isn't new. Most of the files here come from splitting decorated_types.rs into separate files.

File: full-service/src/db/b58.rs [hunk 11]
```diff
@@ -0,0 +1,26 @@
+//! Public Address base58 encoding and decoding.
```
Comment (mfaulk) [inline full-service/src/db/b58.rs:1]: This stuff isn't new, it just moved. Happy to put it somewhere else, but this feels better than leaving it in db/mod.rs

File: full-service/src/json_rpc/mod.rs [hunk 12]
```diff
@@ -0,0 +1,28 @@
...
+mod transaction_log;
+mod txo;
+mod wallet_api;
+mod wallet_status;
+
+pub use account::Account;
+pub use address::Address;
+pub use block::{Block, BlockContents};
+pub use json_rpc_request::Request;
+pub use json_rpc_response::BalanceResponse;
```
Comment (eranrund) [inline full-service/src/json_rpc/mod.rs:20]: Looks like this repo doesn't have a proper rustfmt.toml file, or that rustfmt wasn't ran. We group imports in other places (for example line 27).

File: full-service/src/json_rpc/transaction_log.rs [hunk 13]
```diff
@@ -0,0 +1,71 @@
+// Copyright (c) 2020-2021 MobileCoin Inc.
+
+// use crate::db::models::TransactionLog;
```
Comment (eranrund) [inline full-service/src/json_rpc/transaction_log.rs:3]: Leftovers

File: full-service/src/json_rpc/wallet_api.rs [hunk 14]
```diff
@@ -0,0 +1,877 @@
...
+    command: Json<Request>,
+) -> Result<Json<Response>, String> {
+    let service = service.lock().unwrap();
+
+    let result = match command.0 {
+        Request::create_account { name, first_block } => {
+            let fb = first_block
+                .map(|fb| fb.parse::<u64>())
+                .transpose()
+                .map_err(|e| format!("{{\"error\": \"{:?}\"}}", e))?;
```
Comment (eranrund) [inline full-service/src/json_rpc/wallet_api.rs:38]: I am pretty sure the project now has a better way of doing this... @sugargoat do you remember where you fixed that?

File: full-service/src/json_rpc/wallet_api.rs [hunk 15]
```diff
@@ -0,0 +1,877 @@
...
+        );
+
+        let account_obj = result.get("account").unwrap();
+        assert_eq!(
+            account_obj.get("account_id").unwrap(),
+            expected_response.account.account_id.as_str()
+        );
+    }
+
+    // #[test_with_logger]
```
Comment (eranrund) [inline full-service/src/json_rpc/wallet_api.rs:421]: Is this intentional?

File: full-service/src/service/wallet_service.rs [hunk 16]
```diff
@@ -527,52 +523,57 @@ impl<
...
-            Ok(JsonTx::from(&proto_tx))
+            let proto_tx = mc_api::external::Tx::from(&tx); // TODO: ???
+            Ok(mc_mobilecoind_json::data_types::JsonTx::from(&proto_tx))
         } else {
             Err(WalletServiceError::NoTxInTransaction)
         }
     }
 
+    /*
+
```
Comment (eranrund) [inline full-service/src/service/wallet_service.rs:576]: Leftover?

File: full-service/src/service/wallet_trait.rs [hunk 17]
```diff
@@ -0,0 +1,276 @@
...
+#[automock]
+pub trait Wallet {
+    /// An overview of this wallet.
+    fn get_wallet_status(&self) -> Result<json_rpc::WalletStatus, WalletServiceError>;
+
+    /// Create a new account.
+    ///
+    /// # Arguments
+    /// * `name` - The name of the account.
+    /// * `first_block` - ???
```
Comment (eranrund) [inline full-service/src/service/wallet_trait.rs:19]: My guess is that this is the first block to begin view-key-scanning.

File: full-service/src/service/wallet_trait.rs [hunk 18]
```diff
@@ -0,0 +1,276 @@
...
+
+    /// Import an existing account into this wallet.
+    ///
+    /// # Arguments
+    /// * `entropy` - ???
+    /// * `name` - ???
+    /// * `first_block` - ???
+    fn import_account(
+        &self,
+        entropy: String,
```
Comment (eranrund) [inline full-service/src/service/wallet_trait.rs:34]: My guess: hex encoded 32 bytes from which account keys are derived. It should probably be a &[u8] or &[u8; 32] since the hex encoding is JSON-specific.

File: full-service/src/service/wallet_trait.rs [hunk 19]
```diff
@@ -0,0 +1,276 @@
...
+        name: Option<String>,
+        first_block: Option<u64>,
+    ) -> Result<json_rpc::Account, WalletServiceError>;
+
+    /// Get an account by ID.
+    ///
+    /// # Arguments
+    /// * `account_id_hex` - ???
+    /// * `name` - The new account name.
```
Comment (eranrund) [inline full-service/src/service/wallet_trait.rs:43]: There's no name

File: full-service/src/service/wallet_trait.rs [hunk 20]
```diff
@@ -0,0 +1,276 @@
...
+        &self,
+        entropy: String,
+        name: Option<String>,
+        first_block: Option<u64>,
+    ) -> Result<json_rpc::Account, WalletServiceError>;
+
+    /// Get an account by ID.
+    ///
+    /// # Arguments
+    /// * `account_id_hex` - ???
```
Comment (eranrund) [inline full-service/src/service/wallet_trait.rs:42]: Unique account identifier?

**Review by eranrund — APPROVED** [review]
This is a super nice cleanup!

File: full-service/src/json_rpc/address.rs [hunk 21]
```diff
@@ -0,0 +1,33 @@
...
+#[derive(Deserialize, Serialize, Default, Debug, Clone)]
+pub struct Address {
+    pub object: String,
+    pub address_id: String,
+    pub public_address: String,
+    pub account_id: String,
+    pub address_book_entry_id: Option<String>,
+    pub comment: String,
+    pub subaddress_index: String,
+    pub offset_count: i32,
```
[the remark this answers is not in the document]
Comment (mfaulk) [reply full-service/src/json_rpc/address.rs:15]: Agreed!

File: full-service/src/json_rpc/wallet_api.rs (hunk 15 above)
Comment (mfaulk) [reply full-service/src/json_rpc/wallet_api.rs:421]: Basically, yes. This PR introduces the ability to mock the wallet service, and so the unit tests in this file should pretty much just test that the "plumbing" from JSON, to Rust, into the mock wallet service, and back to the JSON response. The commented-out unit tests below are more like "end-to-end" tests because they actually set up a database and test the wallet service. I'd like to move this kind of testing to the wallet service itself (so that json_rpc isn't involved in testing that service), and / or to crate-level integration tests

Comment (mfaulk) [reply full-service/src/json_rpc/wallet_api.rs:421]: However, maybe it's best to restore these tests now, and deal with them in a follow-up PR? This PR has already gotten too big

File: full-service/src/db/b58.rs [hunk 22]
```diff
@@ -0,0 +1,80 @@
...
+        return Err(WalletDbError::B58Decode);
+    };
+
+    let public_address =
+        PublicAddress::try_from(pubaddr_proto).map_err(|_e| WalletDbError::B58Decode)?;
+    Ok(public_address)
+}
+
+#[cfg(test)]
+mod tests {
```
Comment (mfaulk) [inline full-service/src/db/b58.rs:33]: I sketched out a few unit tests, but I'd like to do those in a separate PR.

File: full-service/src/json_rpc/mod.rs (hunk 12 above)
Comment (mfaulk) [reply full-service/src/json_rpc/mod.rs]: Good catch. I just added a rustsfmt.toml to the project.

File: rustfmt.toml [hunk 23]
```diff
@@ -0,0 +1,5 @@
+merge_imports = true
+
+ignore = [
+    "mobilecoin",   # The mobilecoin submodule is a dependency of this project.
```
Comment (mfaulk) [inline rustfmt.toml:4]: I think ignoring this is the right thing to do? Eventually, this project should probably depend on other mobilecoin crates instead of using a git submodule...

File: full-service/src/service/wallet_trait.rs (hunk 18 above)
Comment (mfaulk) [reply full-service/src/service/wallet_trait.rs:31]: I'd like to leave this for another PR that focuses on getting JSON out of the service layer.

File: full-service/src/json_rpc/address.rs (hunk 21 above)
[the remark this answers is not in the document]
Comment (sugargoat) [reply full-service/src/json_rpc/address.rs:15]: Note - these are currently documented in the [API.md](https://github.com/mobilecoinofficial/full-service/blob/main/API.md#the-assigned-address-object) - it would be nice to move that documentation in with the json_rpc object, and autogenerate something like API.md. API.md was modeled off of stripe's documentation.

File: full-service/src/json_rpc/wallet_api.rs [hunk 24]
```diff
@@ -0,0 +1,880 @@
+// Copyright (c) 2020-2021 MobileCoin Inc.
+
+//! Wallet API.
+//!
+//! A [JSON RPC 2.0](https://www.jsonrpc.org/specification) API inspired by [Stratum](https://github.com/aeternity/protocol/blob/master/STRATUM.md) .
```
Comment (sugargoat) [inline full-service/src/json_rpc/wallet_api.rs:5]: I don't agree that this was inspired by stratum - it was inspired by the [BEAM Wallet API](https://github.com/BeamMW/beam/wiki/Beam-wallet-protocol-API), which references stratum for structure, but BEAM's wallet API was more directly the influence for this API.

File: full-service/src/json_rpc/wallet_api.rs (hunk 14 above)
Comment (sugargoat) [reply full-service/src/json_rpc/wallet_api.rs:42]: It's in the encrypted DB PR, I will be pulling it out as a separate PR - that can happen either before or after this goes in.

File: full-service/src/json_rpc/wallet_api.rs [hunk 25]
```diff
@@ -0,0 +1,880 @@
...
+            .body(body.to_string())
+            .dispatch();
+        assert_eq!(res.status(), Status::Ok);
+        let body = res.body().unwrap().into_string().unwrap();
+        // log::info!(logger, "Attempted dispatch got response {:?}", body);
+        let res: JsonValue = serde_json::from_str(&body).unwrap();
+        res.get("result").unwrap().clone()
+    }
+
+    // fn dispatch_expect_error(
```
Comment (sugargoat) [inline full-service/src/json_rpc/wallet_api.rs:353]: I don't think this should be commented out - are the expected errors not firing?

File: full-service/src/json_rpc/wallet_api.rs [hunk 26]
```diff
@@ -0,0 +1,880 @@
...
+    /// Submit a JSON-RPC request.
+    fn dispatch(client: &Client, body: JsonValue, _logger: &Logger) -> serde_json::Value {
+        let mut res = client
+            .post("/wallet")
+            .header(ContentType::JSON)
+            .body(body.to_string())
+            .dispatch();
+        assert_eq!(res.status(), Status::Ok);
+        let body = res.body().unwrap().into_string().unwrap();
+        // log::info!(logger, "Attempted dispatch got response {:?}", body);
```
Comment (sugargoat) [inline full-service/src/json_rpc/wallet_api.rs:348]:
```suggestion
        log::info!(logger, "Attempted dispatch got response {:?}", body);
```

File: full-service/src/json_rpc/wallet_api.rs (hunk 15 above)
Comment (sugargoat) [reply full-service/src/json_rpc/wallet_api.rs:424]: I disagree with this - these tests were designed to be end-to-end tests, and I would prefer if they stay. We can add additional tests at lower layers, but imo it's important to test the actual rpc endpoints that the API is delivering.

File: full-service/src/service/wallet_trait.rs (hunk 8 above)
Comment (sugargoat) [reply full-service/src/service/wallet_trait.rs:8]: Feel free to look at the [README](https://github.com/mobilecoinofficial/full-service/blob/main/README.md) - many of the ??? fields below are specified there.

**Review by sugargoat — CHANGES_REQUESTED** [review]
Looking good! I would like to see the json_rpc layer tests restored, and for the documentation to be completed before this goes in, but I like the refactor overall.

File: full-service/src/json_rpc/wallet_api.rs (hunk 25 above)
Comment (sugargoat) [reply full-service/src/json_rpc/wallet_api.rs:353]: I see - this was commented out because all the tests were commented out.
